app/src/main/java/com/infomaniak/mail/ui/newMessage/NewMessageAiManager.kt

```diff
@@ -0,0 +1,246 @@
+    private fun onAiPromptOpened(resetPrompt: Boolean) = with(binding) {
+        if (resetPrompt) {
+            aiViewModel.apply {
+                aiPrompt = ""
+                aiPromptOpeningStatus.value?.shouldResetPrompt = false
+            }
+        }
+
+        // Keyboard is opened inside onCreate() of AiPromptFragment
+
+        val foundFragment = childFragmentManager.findFragmentByTag(AI_PROMPT_FRAGMENT_TAG) as? AiPromptFragment
+        if (aiPromptFragment == null) {
+            aiPromptFragment = foundFragment ?: AiPromptFragment()
+        }
+
+        if (foundFragment == null) {
+            childFragmentManager
+                .beginTransaction()
+                .add(aiPromptFragmentContainer.id, aiPromptFragment!!, AI_PROMPT_FRAGMENT_TAG)
+                .commitNow()
+        }
+
+        scrim.apply {
+            isVisible = true
+            isClickable = true
+        }
+
+        animateAiPrompt(true)
+        setAiPromptVisibility(true)
+        newMessageConstraintLayout.descendantFocusability = FOCUS_BLOCK_DESCENDANTS
+    }
+
+    private fun onAiPromptClosed(withoutTransition: Boolean) = with(binding) {
+
+        fun removeFragmentAndHideScrim() {
+            aiPromptFragment?.let {
+                childFragmentManager
+                    .beginTransaction()
+                    .remove(it)
+                    .commitNow()
+            }
+            aiPromptFragment = null
+
+            scrim.isGone = true
+        }
+
+        aiPromptFragmentContainer.hideKeyboard()
+
+        if (withoutTransition) {
+            removeFragmentAndHideScrim()
+        } else {
+            viewLifecycleOwner.lifecycleScope.launch {
+                delay(animationDuration)
+                removeFragmentAndHideScrim()
+            }
+        }
+
+        if (!withoutTransition) animateAiPrompt(false)
+        setAiPromptVisibility(false)
+        newMessageConstraintLayout.descendantFocusability = FOCUS_BEFORE_DESCENDANTS
+    }
+
+    private fun animateAiPrompt(isVisible: Boolean) = with(binding) {
+        val slidingTransition = Slide()
+            .addTarget(aiPromptLayout)
+            .setDuration(animationDuration)
+
+        TransitionManager.beginDelayedTransition(root, slidingTransition)
+
+        val (startOpacity, endOpacity) = if (isVisible) 0.0f to scrimOpacity else scrimOpacity to 0.0f
+
+        valueAnimator?.cancel()
+        valueAnimator = ValueAnimator.ofObject(FloatEvaluator(), startOpacity, endOpacity).apply {
+            duration = animationDuration
+            addUpdateListener { animator ->
+                val alpha = ((animator.animatedValue as Float) * 256.0f).roundToInt() / 256.0f
+                scrim.alpha = alpha
+                activity.window.statusBarColor = UiUtils.pointBetweenColors(backgroundColor, black, alpha)
+            }
+            start()
+        }
+    }
+
+    private fun setAiPromptVisibility(isVisible: Boolean) {
+
+        fun updateNavigationBarColor() {
```

Comment:
Any reasons to add that `inner function` here ? Since you're only using this once.

Reply:
Sometimes we took the habit of creating some inner functions (in opposition to class methods) here and there just so they stay closer to where they are used. No better reason.

Making these two lines into a function is to group things together more neatly though I'd say